src/bsp/src/esp/src/SpiStm.cpp

```diff
@@ -0,0 +1,198 @@
+            m_logger.log(LogLevel::Debug, "Bytes were: | %d | %d | %d | %d |",
+                         m_inboundMessage.m_data[0], m_inboundMessage.m_data[1],
+                         m_inboundMessage.m_data[2], m_inboundMessage.m_data[3]);
+            m_rxState = receiveState::ERROR;
+            break;
+        }
+
+        if (WORDS_TO_BYTES(m_inboundHeader->rxSizeWord) == m_outboundMessage.m_sizeBytes &&
+            m_outboundMessage.m_sizeBytes != 0) {
+            m_logger.log(LogLevel::Debug, "Received valid header. Can now send payload");
+            gpio_set_level(STM_USER_0, 0);
+            m_txState = transmitState::SENDING_PAYLOAD;
+        } else {
+            m_txState = transmitState::SENDING_HEADER;
+            m_logger.log(LogLevel::Debug, "Received valid header but cannot send payload");
+        }
+
+        // This will be sent on next header. Payload has priority over headers.
+        m_inboundMessage.m_sizeBytes = WORDS_TO_BYTES(m_inboundHeader->txSizeWord);
+        if (m_inboundMessage.m_sizeBytes == WORDS_TO_BYTES(m_outboundHeader.rxSizeWord) &&
+            m_inboundMessage.m_sizeBytes != 0) {
+            rxLengthBytes = WORDS_TO_BYTES(m_inboundHeader->txSizeWord);
+            m_logger.log(LogLevel::Debug, "Receiving payload");
+            m_rxState = receiveState::RECEIVING_PAYLOAD;
+        } else {
+            rxLengthBytes = StmHeader::sizeBytes;
+            m_rxState = receiveState::RECEIVING_HEADER;
+        }
+        break;
+    case receiveState::RECEIVING_PAYLOAD:
+        rxLengthBytes = WORDS_TO_BYTES(m_inboundHeader->txSizeWord);
+        break;
+    case receiveState::VALIDATE_CRC:
+        if (calculateCRC32_software(m_inboundMessage.m_data.data(),
+                                    m_inboundMessage.m_sizeBytes - CRC32_SIZE) !=
+            *(uint32_t*)&m_inboundMessage.m_data[m_inboundMessage.m_sizeBytes - CRC32_SIZE]) {
+            m_outboundHeader.systemState.stmSystemState.failedCrc = 1;
+        } else {
+            // TODO: Replace with proper reception handle
+            m_logger.log(LogLevel::Info, "Stm says: %s", m_inboundMessage.m_data.data());
+            m_inboundMessage.m_sizeBytes = 0;
+        }
+        m_rxState = receiveState::RECEIVING_HEADER;
+        rxLengthBytes = StmHeader::sizeBytes;
+        break;
+    case receiveState::ERROR:
+        m_logger.log(LogLevel::Error, "Error within Spi driver STM");
+        m_rxState = receiveState::RECEIVING_HEADER;
+        break;
+    }
+
+    switch (m_txState) {
+    case transmitState::SENDING_HEADER:
+        updateOutboundHeader();
+        m_transaction.tx_buffer = &m_outboundHeader;
+        txLengthBytes = StmHeader::sizeBytes;
+        break;
+    case transmitState::SENDING_PAYLOAD:
+        m_transaction.tx_buffer = m_outboundMessage.m_data.data();
+        txLengthBytes = m_outboundMessage.m_sizeBytes;
+        break;
+    case transmitState::ERROR:
+        break;
```

Comment:
Log if there is an error, unless the error is frequent, but I don't think it's the case here